core/src/main/java/org/infinispan/remoting/inboundhandler/GlobalInboundInvocationHandler.java

```diff
@@ -135,54 +126,61 @@ private void initializeCacheRpcCommand(CacheRpcCommand command, ComponentRegistr
+         reply.reply(exceptionHandlingCommand(throwable));
+      }
+   }
+
+   private void handleReplicableCommand(Address origin, ReplicableCommand command, Reply reply, DeliverOrder order) {
+      if (trace) {
+         log.tracef("Attempting to execute non-CacheRpcCommand: %s [sender=%s]", command, origin);
+      }
+      if (order.preserveOrder() || !command.canBlock()) {
+         runReplicableCommand(command, reply, order.preserveOrder());
+      } else {
+         remoteCommandsExecutor.execute(() -> runReplicableCommand(command, reply, order.preserveOrder()));
+      }
+   }
+
+   private void runReplicableCommand(ReplicableCommand command, Reply reply, boolean preserveOrder) {
+      try {
+         invokeReplicableCommand(command, reply, preserveOrder);
+      } catch (Throwable throwable) {
+         if (throwable.getCause() != null && throwable instanceof CompletionException) {
+            throwable = throwable.getCause();
+         }
+         if (throwable instanceof InterruptedException || throwable instanceof IllegalLifecycleStateException) {
+            log.shutdownHandlingCommand(command);
+            reply.reply(shuttingDownResponse());
+         } else {
+            log.exceptionHandlingCommand(command, throwable);
+            reply.reply(exceptionHandlingCommand(throwable));
          }
-      };
+      }
    }
 
-   private Runnable create(final ReplicableCommand command, final Reply reply, boolean preserveOrder) {
-      return new Runnable() {
-         @Override
-         public void run() {
-            try {
-               globalComponentRegistry.wireDependencies(command);
-
-               CompletableFuture<Object> future = command.invokeAsync();
-               if (preserveOrder) {
-                  future.join();
-               } else {
-                  future.whenComplete((retVal, throwable) -> {
-                     if (retVal != null && !(retVal instanceof Response)) {
-                        retVal = SuccessfulResponse.create(retVal);
-                     }
-                     reply.reply(retVal);
-                  });
-               }
-            } catch (Throwable throwable) {
-               if (throwable.getCause() != null && throwable instanceof CompletionException) {
-                  throwable = throwable.getCause();
-               }
-               if (throwable instanceof InterruptedException || throwable instanceof IllegalLifecycleStateException) {
-                  log.shutdownHandlingCommand(command);
-                  reply.reply(shuttingDownResponse());
-               } else {
-                  log.exceptionHandlingCommand(command, throwable);
-                  reply.reply(exceptionHandlingCommand(throwable));
-               }
+   private void invokeReplicableCommand(ReplicableCommand command, Reply reply, boolean preserveOrder)
+         throws Throwable {
+      globalComponentRegistry.wireDependencies(command);
```

Comment:
`@Inject` in `CacheTopologyControlCommand` is convenient, but not essential, so we can do things manually if it shaves 0.1% off CPU time. Another PR, though :)

I'm fine with the ack collector in the externalizer, I definitely wouldn't want to use reflection to set it on the ack commands.